components/util/model.jsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { immutableMerge } from './index.js';
+
+export function Model({ initialState, methods }) {
+  const listeners = [];
+  const notifyListeners = nextState => listeners.forEach(fn => fn(nextState));
+
+  const model = {
+    state: Object.freeze(initialState),
+    subscribe: fn => listeners.push(fn),
+    unsubscribe: fn => listeners.splice(listeners.indexOf(fn), 1),
+  };
+
+  Object.keys(methods).forEach((method) => {
+    model[method] = (...args) => {
+      const updates = methods[method].apply(model, args);
+      if (updates && typeof updates === 'object') {
```

Comment:
I wanted state to be allowed to be an array or an object initially, but since I am using `Object.assign`, if you pass it an array you end up with state that looks like `{0: 'foo', 1: 'bar', 2: 'baz' }` (and then none of the array methods would work).

So... I should explicitly not allow array state. Good point.